### PR TITLE
[FIXED JENKINS-37041] Ensure that detached plugins are always at least their minimum version

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-37041](https://issues.jenkins-ci.org/browse/JENKINS-37041)

Ok, so here was the test scenario:
1. Upgrade Jenkins (this will upgrade the detached plugin)
2. Stop Jenkins
3. Overwrite the detached plugin with the older version
4. Start Jenkins
5. Watch as the world burns

With this fix, under my _(slightly contrived)_ test scenario I now get

```
INFO: Started initialization
Jul 28, 2016 4:07:38 PM hudson.PluginManager loadDetachedPlugins
WARNING: Detached plugin bouncycastle-api found at version 1.648.3, required minimum version is 2.16.0
...
INFO: Initializing Bouncy Castle security provider.
Jul 28, 2016 4:07:39 PM jenkins.bouncycastle.api.SecurityProviderInitializer addSecurityProvider
INFO: Bouncy Castle security provider initialized.
Jul 28, 2016 4:07:42 PM jenkins.InitReactorRunner$1 onAttained
```

So this fixed the only code path that I could find that remained whereby a plugin that has been detached would not get upgraded (namely if the admin mistakenly overwrites the detached plugin with an older version than the stated minimum version)

CRITICAL IMHO that this lands in 2.16

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/ClassicPluginStrategy.java [hunk 1]
```diff
@@ -401,6 +401,16 @@ public String getShortName() {
...
             return splitWhen;
         }
+
+        /**
+         * Gets the minimum required version for the current version of Jenkins.
+         *
+         * @return the minimum required version for the current version of Jenkins.
+         * @sice 2.16
+         */
+        public VersionNumber getRequireVersion() {
```
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: :bug: rename to `getRequiredVersion` (and rename field & constructor parameter to match)

**Comment by daniel-beck** [comment]
Isn't this usually something that usually shows up in the detached plugin admin monitor?

Not opposed, just wondering how they relate.

File: core/src/main/java/hudson/PluginManager.java [hunk 2]
```diff
@@ -716,6 +716,34 @@ public boolean accept(File dir, String name) {
                     new Object[] {lastExecVersion, Jenkins.VERSION, loadedDetached});
 
             InstallUtil.saveLastExecVersion();
+        } else {
+            final Set<ClassicPluginStrategy.DetachedPlugin> forceUpgrade = new HashSet<>();
+            for (ClassicPluginStrategy.DetachedPlugin p : ClassicPluginStrategy.getDetachedPlugins()) {
+                VersionNumber installedVersion = getPluginVersion(rootDir, p.getShortName());
+                VersionNumber requireVersion = p.getRequireVersion();
```
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java]: :ant: ditto here

**Comment by jglick** [comment]
Argh

```
npm ERR! shasum check failed for /tmp/npm-1725-b4cb96c5/registry.npmjs.org/jshint/-/jshint-2.9.2.tgz
```

**Comment by jglick** [comment]
:bee:

**Comment by stephenc** [comment]
@daniel-beck so I believe the scenario that I have fixed addresses the root issue that @oleg-nenashev was trying to fix in https://github.com/jenkinsci/jenkins/pull/2436 (though that particular PR seems overly heavyweight and complex to fix this problem)

In the specific case of 2.16... you will be in a world of pain if bouncycastle does not load and hence you may not (depending on the other plugins you have installed) be even able to see the admin monitor... given that we have the plugin in the detached list and can therefore fix things for the user before unpacking the plugin archives, seems only prudent to fix things for them

**Comment by abayer** [comment]
:bee: assuming tests pass.

**Comment by jglick** [comment]
re:bee:

**Comment by stephenc** [comment]
@reviewbybees done

**Comment by stephenc** [comment]
I'm going to wait for the tests... @jenkinsci/code-reviewers as I view this fix as critical, anyone who wants to object has until the tests pass before I hit merge

**Comment by stephenc** [comment]
@jenkinsci/code-reviewers if you want extra time to review, just say so and I will wait... but if I see the tests have passed and nobody else has either objected or asked for more time to review, then I will hit merge ;-)

**Comment by stephenc** [comment]
Ok tests passed

**Comment by oleg-nenashev** [comment]
5 hours timeout is definitely not enough for @jenkinsci/code-reviewers :sadpanda:

**Comment by oleg-nenashev** [comment]
The code looks good to me, but I'm not so happy about merging such higly-important changes without any tests

**Comment by daniel-beck** [comment]
This may have caused [JENKINS-37332](https://issues.jenkins-ci.org/browse/JENKINS-37332).

**Pull request merged**